chain-signatures/contract/src/lib.rs

```diff
@@ -178,46 +216,101 @@ impl VersionedMpcContract {
-            predecessor,
-            payload,
-            path,
-            key_version
-        );
 
+        let predecessor = env::predecessor_account_id();
         let request = SignatureRequest::new(payload, &predecessor, &path);
-        match self.sign_result(&request) {
-            None => {
-                self.add_sign_request(&request);
-                env::log_str(&serde_json::to_string(&near_sdk::env::random_seed_array()).unwrap());
-                Self::ext(env::current_account_id()).sign_helper(request, 0)
+        if !self.request_already_exists(&request) {
+            match self {
+                Self::V0(mpc_contract) => {
+                    let index = mpc_contract.next_available_yield_resume_request_index;
+                    mpc_contract.next_available_yield_resume_request_index += 1;
+
+                    let yield_promise = env::promise_yield_create(
+                        "clear_state_on_finish",
+                        &serde_json::to_vec(&(index,)).unwrap(),
+                        CLEAR_STATE_ON_FINISH_CALL_GAS,
+                        GasWeight(0),
+                        DATA_ID_REGISTER,
+                    );
+
+                    // Store the request in the contract's local state
+                    let data_id: CryptoHash = env::read_register(DATA_ID_REGISTER)
+                        .expect("")
+                        .try_into()
+                        .expect("");
+
+                    mpc_contract.add_request(&request, &Some(data_id));
+                    mpc_contract.add_yield_resume_request(
+                        index,
+                        YieldResumeRequest {
+                            data_id,
+                            account_id: env::signer_account_id(),
+                            signature_request: request,
+                        },
+                    );
+
+                    log!(
+                        "sign: predecessor={}, payload={:?}, path={:?}, key_version={}, data_id={:?}",
+                        predecessor,
+                        payload,
+                        path,
+                        key_version,
+                        data_id
+                    );
+                    env::log_str(
+                        &serde_json::to_string(&near_sdk::env::random_seed_array()).unwrap(),
+                    );
+
+                    let final_yield_promise = env::promise_then(
+                        yield_promise,
+                        env::current_account_id(),
+                        "return_signature_on_finish",
+                        &[],
+                        NearToken::from_near(0),
+                        RETURN_SIGNATURE_ON_FINISH_CALL_GAS,
+                    );
+                    // The return value for this function call will be the value
+                    // returned by the `sign_on_finish` callback.
+                    env::promise_return(final_yield_promise);
```

Comment:
Seems like this promise is just unwrapping the results which isn't really needed. We can get rid of this by just directly handling the result in the `clear_state_on_finish` function

Reply:
this is needed because if we do handle_result in the clear_state_on_finish function, then when it panics, the state changes (remove pending requests) will be reverted and then cause `tx finished unsuccessfully: Failure(ActionError(ActionError { index: Some(0), kind: FunctionCallError(ExecutionError("Smart contract panicked: Signature for this payload already requested")) }))` in the `test_signature_offline_node_back_online` test.

Reply:
I see, alright ignore this one then